solr/core/src/java/org/apache/solr/core/SolrCore.java

```diff
@@ -487,10 +488,13 @@ public String getName() {
   }
 
   public void setName(String v) {
+    Objects.requireNonNull(v);
+    boolean renamed = this.name != null && !this.name.equals(v);
+    assert !renamed || coreDescriptor.getCloudDescriptor() == null : "Cores are not renamed in SolrCloud";
```

Comment:
Should it call apiAssumeStandalone()?

Reply:
I chose to name "apiAssumeStandalone" that because it's intended for primary service/API methods (likely coming from a user) yielding a BAD_REQUEST instead of SERVER_ERROR.  SolrCore.setName is more of an internal detail method where an assertion seemed best.  Perhaps I should add a method on SolrCore "isSolrCloud" or "isStandalone" and/or assumeStandalone or assumeSolrCloud and use these more pervasively?  I feel that could be its own issue since there are many places that could be enhanced/clarified to use this.  Across Solr there are various techniques like checking the subclass of the core's SolrResourceLoader or checking if the CoreContainer has a ZkController.  A consistent approach would be nicer; perhaps separate for CoreContainer level vs SolrCore level

Reply:
+1 to have a consistent and more pervasive approach. It would first make the code clearer to use, and would probably avoid bugs.

Reply:
I'll file a new issue for that.